### ISSUE TITLE
Syncing of Audio and Video with Custom Listener

### ISSUE THREAD

**Issue by dirtygilly** [RenderHeads/UnityPlugin-AVProMovieCapture#423]
Hello

I have build a custom audio "Capture from Audio Listener Component" script which inherits from UnityAudioCapture class, to incorporate two audio sources (microphone and screen) based on this post https://github.com/RenderHeads/UnityPlugin-AVProMovieCapture/issues/96 and this code https://gist.github.com/WikkidEdd/35fea92260b95135d6837188006919dc. 

It works great! However I get a creeping audio-video synch. I thought it may be my custom plugin processing time but the audio runs ahead of the video. 

Do you have any suggestions on how to address or trouble shoot the problem?

Thanks!

**Comment by Chris-RH** [comment]
We don't officially support capturing audio from multiple sources, and that workaround is something that a user coded for themselves, so its not really anything to do with us.  
That said, could you provide more info and maybe a basic test project to unitysupport@renderheads.com, then there might be something that we can do.  A sample of the video recorded would also be helpful, thanks.

1. Unity Version
2. AVPro Movie Capture Version
3. Which platform(s) are you using?
4. Which OS version(s) are you using?
5. Which rendering API(s) are you using?
6. Hardware
7. Which capture component are you using?
8. Capture mode
9. Which output mode are you using?
10. Video codecs
11. Audio source
12. Audio codecs
13. Any other component configuration
14. Log output

**Comment by dirtygilly** [comment]
After giving it some thought, I think it is a bit rate misalignment between the video and audio, given the audio runs ahead  i.e. too few audio samples streamed. Switching between 2 sources is probably causing or making it worse. I am going to trouble shoot down this path a bit more and will reopen this ticket, if needs be. If you do think of something or have suggestions, that would be helpful as well. Thanks for your prompt responses.

**Comment by hyper-active** [comment]
@dirtygilly That's really cool, did you make any progress on this or care to share your current script? I've been trying to find a solution to this for a while, wanting to record both game audio and microphone.

**Comment by dirtygilly** [comment]
> [@dirtygilly](https://github.com/dirtygilly) That's really cool, did you make any progress on this or care to share your current script? I've been trying to find a solution to this for a while, wanting to record both game audio and microphone.

Yes I eventually got it working. I created a "Capture Audio From Multiple Sources" (CAMS) script based on what I learnt in the script in the link in my initial post above. In summary the solution was comprised:
1) Two separate game objects with:
    a) A RenderHeads Capture from Screen script (audio source set to "Unity") as well as my CAMS script    
    b) A RenderHeads Capture from WebcamTexture (audio source set to "Unity") as well as my CAMS script    
2) On the main camera I have an "On Audio Filter Read" (OAFR) script
3) I created my own script to control the Microphone and capture the audio data stream from that

The scripts in 2 and 3 basically capture the audio streams from the screen and the webcam respectively. This is sent to the CAMS script which figures out which audio stream to use to pass to the RenderHeads script for embedding in the video. My audio stream capture was discrete (either microphone or screen) but I remember it being able todo both.

My final issue was that the audio would become out of synch with the video after like 30 seconds. I eventually solved that by elevating the execution order of the CAMS and the OAFR script to run quite high in the priority list and before the RenderHeads script.

Hope this helps?
